Cram: sanitize error messages

rgrinberg opened the pull request:
Fix #3673

rgrinberg commented:
Hmm, I'm unable to reproduce the error in CI. @emillon mind trying this out? (if you use Linux)

rgrinberg commented on src/dune/cram_exec.ml:
```diff
@@ -74,7 +80,7 @@ end = struct
       | Ok None -> ()
       | Ok (Some sexp) ->
-        let command = Command.of_sexp sexp in
+        let command = Command.of_sexp (assert false) sexp in
```
This is dead code anyway

jeremiedimino commented:
The printing with `git diff` in the CI doesn't look right BTW.

rgrinberg commented:
> The printing with git diff in the CI doesn't look right BTW.

Unfortunately, I'm unable to reproduce the issue locally. I assume this is a difference of behavior in the shells that are being used. Do you think it's worth delaying the release for this? I understand that it would be better to have the same behavior since the beginning, but this doesn't seem possible in this case if we're willing to keep updating the regular expressions for different shells.

jeremiedimino commented:
That seems fine. We'll just need to tie the regexp to the version of the dune language, which seems fine.

Regarding this PR, when I look at the CI failure, there seems to be two issues:
- there is still a `: ` prefix (I can reproduce it on my machine)
- dune calls `git diff --no-index ...` as diff program and instead of printing a diff between `run.t` and `run.t.corrected`, it shows a diff between `run.t` and `/dev/null` and then a diff between `/dev/null` and `run.t.corrected`

The second problem seems independent of this PR, but it'd be good to look into it at some point. When patdiff is not installed, dune calls `git diff` by default and this double diffing behaviour is a pretty bad user experience.

rgrinberg commented:
> there is still a : prefix (I can reproduce it on my machine)

I believe I fixed that.

> dune calls git diff --no-index ... as diff program and instead of printing a diff between run.t and run.t.corrected, it shows a diff between run.t and /dev/null and then a diff between /dev/null and run.t.corrected

That's not exactly what's happening. The diff command seems to be correct, but there seems to be a weird interaction between promotion, sandboxing, and git's diff that's causing the issue. AFAIK, the way promotion works is as follows:

* We run diff (git diff in this case) against the two files.
* If diff detects a difference, we move the corrected file into the promotion staging area.
* We register the file from the staging area to be promoted.

As far as I can tell, there seems to be some sort of a race and `git diff` is observing the file after it was moved to the staging area. I have no idea why it's happening because `git diff` should finish running before we move the file to staging area, but this is the problem from what I've diagnosed.

In any case, this is indeed unrelated to this PR and I've tweaked the error message to match Linux. I've disabled the test on MacOS as I'm getting a different error message.

jeremiedimino commented:
The regexp should be anchored, but otherwise looks good to merge.

rgrinberg commented on src/dune/cram_exec.ml:
```diff
@@ -136,11 +142,12 @@ let run_expect_test file ~f =
     Build_path_prefix_map.encode_map
       [ Some
           { source = Path.to_absolute_filename cwd; target = "$TESTCASE_ROOT" }
+      ; Some { source = Path.to_absolute_filename temp_dir; target = "$TMPDIR" }
```
document this variable being sanitized.
